CLAUDE.md file for Claude Code (#89)
solana-foundation/solana-attestation-service: CLAUDE.md

## Common Development Commands

### Building and Testing

```bash
# Build the Solana program
cargo-build-sbf

# Run integration tests
cargo-build-sbf && SBF_OUT_DIR=$(pwd)/target/sbf-solana-solana/release cargo test

# Run TypeScript client tests
cd clients/typescript && npm test
```

### IDL Generation

```bash
# Generate IDL using Shank
shank idl -r program -o idl

# Or use the npm script
pnpm run generate-idl
```

### Client Generation

```bash
# Generate Rust and TypeScript clients from IDL
pnpm run generate-clients
```

### TypeScript Client Development

```bash
# Build TypeScript client
cd clients/typescript && npm run build

# Run tests
cd clients/typescript && npm test
```

## Architecture Overview

### Core Components

The system consists of three main components:

1. **Credential**: Represents an issuer's identity and authority to create schemas and attestations
   - Contains issuer name and list of authorized signers
   - Controls who can create schemas and attestations

2. **Schema**: Defines the structure and metadata for attestations
   - Includes field definitions, layout, and validation rules
   - Can be paused/resumed and versioned

>    - Supports tokenization for creating NFT-like attestations

Reviewer: ```suggestion
   - Supports tokenization for creating NFT attestations utilizing Token 2022
```

  - **rust/src/generated/**: Rust client (accounts, instructions, types, errors)
  - **typescript/src/generated/**: TypeScript client (accounts, instructions, programs, types)
  - **typescript/src/**: Client utilities (`pdas.ts`, `utils.ts`) and tests

- **integration_tests/**: Comprehensive test suite for all program functionality
  - **tests/helpers/**: Test utility functions and program context setup
  - **tests/**: Individual test files for each instruction type and tokenization

- **scripts/**: Build and generation scripts
  - Client generation (`generate-clients.js`) and IDL processing (`process-idl.js`)

- **examples/**: Usage examples and practical demonstrations
  - **typescript/**: TypeScript examples for credential setup, schemas, and attestations

- **idl/**: Interface Definition Language files
  - **solana_attestation_service.json**: Program IDL defining accounts, instructions, and types

### Important Constants

- Program ID: `22zoJMtdu4tQc2PzL74ZUT7FrwgB1Udec8DdW4yw4BdG`
- Token Program: `TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb` (Token-2022)
- Event Authority PDA: `DzSpKpST2TSyrxokMXchFz3G2yn5WEGoxzpGEUDjCX4g`

## Technology Stack

### Program Framework

> The program uses **Pinocchio** instead of Anchor, chosen for:

Reviewer: ```suggestion
The program uses [**Pinocchio**](https://github.com/anza-xyz/pinocchio) instead of Anchor, chosen for:
```

- Smaller program size and reduced compute costs
- More control over instruction parsing and execution
- Direct account validation and manipulation

### IDL and Client Generation

> - **Shank**: Used for IDL generation with `#[derive(ShankInstruction)]` and `#[account()]` attributes
> - **Codama**: Transforms Shank-generated IDL into TypeScript and Rust clients

Reviewer: ```suggestion
- [**Shank**](https://github.com/metaplex-foundation/shank): Used for IDL generation with `#[derive(ShankInstruction)]` and `#[account()]` attributes
- [**Codama**](https://github.com/codama-idl/codama): Transforms Shank-generated IDL into TypeScript and Rust clients
```

- Reuse verification functions from `program/src/processor/shared/` for consistency
- Comprehensive PDA validation for all accounts
- Verify program ownership and account derivation
- Check account mutability and signer requirements

## Development Guidelines

### Instruction Implementation

When implementing new instructions:

1. Add instruction variant to `AttestationServiceInstruction` enum in `instructions.rs`
2. Update discriminator matching in `entrypoint.rs`
3. Implement processor function in `processor/` directory with `#[inline(always)]`
4. Follow processor code structure:
   - Account destructuring using array pattern matching
   - Verification and security checks using shared functions from `processor/shared/`
   - Business logic implementation
   - Account resizing if required (implement performantly with proper calculations)
   - Event emission if required
5. Use PDA utilities from `processor/shared/pda_utils.rs` for account derivation
6. Add comprehensive tests in `integration_tests/`
7. Regenerate clients with `pnpm run generate-clients`

### Schema Design

Schema layouts use numeric type identifiers:

> - Primitive types: 0-11 (u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, bool, char)
> - String: 12
> - Vectors: 13-25 (Vec<primitive_type>, Vec<String>)

Reviewer: Might as well use this from our [docs](https://github.com/solana-foundation/solana-attestation-site/blob/development/docs/pages/schemas/index.mdx). 
```suggestion
### Schema Layout Data Types

The `layout` field uses numeric values to specify data types for each field in the schema:

| Value | Data Type | Description |
|-------|-----------|-------------|
| 0 | U8 | Unsigned 8-bit integer |
| 1 | U16 | Unsigned 16-bit integer |
| 2 | U32 | Unsigned 32-bit integer |
| 3 | U64 | Unsigned 64-bit integer |
| 4 | U128 | Unsigned 128-bit integer |
| 5 | I8 | Signed 8-bit integer |
| 6 | I16 | Signed 16-bit integer |
| 7 | I32 | Signed 32-bit integer |
| 8 | I64 | Signed 64-bit integer |
| 9 | I128 | Signed 128-bit integer |
| 10 | Bool | Boolean value |
| 11 | Char | Single character |
| 12 | String | Variable-length string |
| 13 | VecU8 | Vector of unsigned 8-bit integers |
| 14 | VecU16 | Vector of unsigned 16-bit integers |
| 15 | VecU32 | Vector of unsigned 32-bit integers |
| 16 | VecU64 | Vector of unsigned 64-bit integers |
| 17 | VecU128 | Vector of unsigned 128-bit integers |
| 18 | VecI8 | Vector of signed 8-bit integers |
| 19 | VecI16 | Vector of signed 16-bit integers |
| 20 | VecI32 | Vector of signed 32-bit integers |
| 21 | VecI64 | Vector of signed 64-bit integers |
| 22 | VecI128 | Vector of signed 128-bit integers |
| 23 | VecBool | Vector of boolean values |
| 24 | VecChar | Vector of characters |
| 25 | VecString | Vector of strings |

For example, a layout of `[12, 0, 12]` would define three fields: a String, followed by a U8 integer, followed by another String. The `fieldNames` array provides human-readable names that correspond positionally to each layout value, so with field names `["name", "age", "country"]`, the first String field would be named "name", the U8 field would be "age", and the second String field would be "country". Here's an example of how those would be utilized when creating a Schema:

```

### Tokenization Flow

For tokenized attestations:

> 1. Create regular attestation first
> 2. Initialize Token-2022 mint with required extensions
> 3. Set up metadata and group membership
> 4. Create associated token account for recipient
> 5. Mint exactly 1 token to recipient

Reviewer: this should
```suggestion
1. Create regular Credential and Schema
2. Tokenize Schema
3. Utilize Create Tokenize Attestation instruction (this will create the Attestation and Token)
```

### Testing Practices

- Test both success and failure cases
- Use helper functions for common setup operations
- Verify account states before and after operations
- Test edge cases like expired attestations and invalid signers

### Client Integration

- TypeScript clients provide typed interfaces for all instructions
- Rust clients use `solana_program` types for compatibility
- Both clients handle account resolution and instruction building

## Configuration & Environment

### Program Constants

Key seeds and identifiers are defined in `program/src/constants.rs`:
- `CREDENTIAL_SEED`: For credential PDA derivation
- `SCHEMA_SEED`: For schema PDA derivation
- `ATTESTATION_SEED`: For attestation PDA derivation
- `SCHEMA_MINT_SEED`: For tokenized schema mints

> - `ATTESTATION_MINT_SEED`: For tokenized attestation mints

Reviewer: ```suggestion
- `ATTESTATION_MINT_SEED`: For tokenized attestation mints
- `EVENT_AUTHORITY_SEED`: For event authority PDA
-  `SAS_SEED`: For SAS Program Authority PDA;
```

   - Assert that all program accounts are owned by the correct program
   - Verify system accounts are owned by System Program
   - Check token accounts are owned by Token Program

2. **Signer Verification**:
   - Assert required accounts are signers using `account.is_signer()`
   - Verify authorized signers against credential's signer list
   - Validate authority accounts have signing privileges

3. **Account Mutability Checks**:
   - Assert writable accounts using `account.is_writable()`
   - Verify rent-exempt status for account modifications

4. **Discriminator Validation**:
   - Assert account discriminators match expected types (Credential, Schema, Attestation)
   - Prevent type confusion attacks by validating account data structure
   - Check account initialization state

5. **PDA Derivation Verification**:
   - Assert PDA accounts match expected derivation using seeds
   - Verify bump seeds to prevent canonical bump attacks
   - Validate account addresses against program-derived addresses

### Access Control

- All operations require authorized signer verification against credential's authorized signers list
- Credential owners control schema creation and management 

> - Only authorized signers can create, modify, and close attestations

Reviewer: I don't think we have any modify instructions atm.

```suggestion
- Only authorized signers can create and close attestations
```
